server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,153 @@
+ * This class does throttling on task submission to master, it uses class name of request of tasks as key for
+ * throttling. Throttling will be performed over task executor's class level, different task types have different executors class.
+ *
+ * Set specific setting to for setting the threshold of throttling of particular task type.
+ * e.g : Set "master.throttling.thresholds.put_mapping" to set throttling limit of "put mapping" tasks,
+ *       Set it to default value(-1) to disable the throttling for this task type.
+ */
+public class MasterTaskThrottler implements TaskBatcherListener {
+    private static final Logger logger = LogManager.getLogger(MasterTaskThrottler.class);
+
+    public static final Setting<Settings> THRESHOLD_SETTINGS = Setting.groupSetting(
+        "master.throttling.thresholds.",
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * To configure more task for throttling, override getMasterThrottlingKey method with task name in task executor.
+     * Verify that throttled tasks would be retry.
+     *
+     * Added retry mechanism in TransportMasterNodeAction so it would be retried for customer generated tasks.
+     */
+    public static Set<String> CONFIGURED_TASK_FOR_THROTTLING = Collections.unmodifiableSet(
+        // TODO
+        // Add throttling key for other master task as well.
+        // Will be added as part of upcoming PRs,
+        // Right now just added "put-mapping" for ref only.
+        new HashSet<>(Arrays.asList("put-mapping"))
+    );
+
+    private final int DEFAULT_THRESHOLD_VALUE = -1; // Disabled throttling
+    private final MasterService masterService;
+    private final MasterTaskThrottlerListener masterTaskThrottlerListener;
+
+    private final ConcurrentMap<String, Long> tasksCount;
+    private final ConcurrentMap<String, Long> tasksThreshold;
+
+    public MasterTaskThrottler(
+        final ClusterSettings clusterSettings,
+        final MasterService masterService,
+        final MasterTaskThrottlerListener masterTaskThrottlerListener
+    ) {
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.masterService = masterService;
+        this.masterTaskThrottlerListener = masterTaskThrottlerListener;
+        tasksCount = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+        tasksThreshold = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+    }
+
+    public void validateSetting(final Settings settings) {
+        /**
+         * TODO: Change the version number of check as per version in which this change will be merged.
+         */
+        if (masterService.state().nodes().getMinNodeVersion().compareTo(Version.V_3_0_0) < 0) {
+            throw new IllegalArgumentException("All the nodes in cluster should be on version later than or equal to 3.0.0");
```

Comment:
Why is there are restriction on having all node above 3.0?

Reply:
We want to make it to the version to which PR is going to be merged, as of now kept it as Current Version(i.e 3.0).
I have added TODO for changing the version as well. 

Once we raise final PR against master branch will change the version in it and take care of TODO.